src/game/client/water/w3dwatertracks.cpp

```diff
@@ -13,15 +13,324 @@
+    m_totalMs = m_distanceFromShore / m_velocity;
+    m_fadeMs = s_waveTypeInfo[m_type].time_to_fade;
+    m_widthFraction = length * s_waveTypeInfo[m_type].initial_width_fraction;
+    m_heightFraction = m_widthFraction * s_waveTypeInfo[m_type].initial_height_fraction;
+    m_length = length;
+    m_width = width;
+
+    m_timingUnk1 = (m_distanceFromShore - m_width) / m_velocity;
+    m_velocityUnk = -(m_velocity * m_velocity) / (2.0f * m_width);
+    m_timingUnk2 = -m_velocity / m_velocityUnk;
+    m_timingUnk3 = GameMath::Sqrt(GameMath::Fabs(2.0f * m_width / m_velocityUnk));
+    m_totalMs = (int)(m_timingUnk1 + m_timingUnk2 + m_timingUnk3);
+    m_scaleUnk = 2.0f * m_heightFraction / (m_timingUnk2 * m_timingUnk2);
+    m_timeToCompress = s_waveTypeInfo[m_type].time_to_compress;
+
+    if (m_type == 5) {
+        m_timingUnk3 = 1000.0f;
+        m_totalMs = (int)(m_timingUnk1 + m_timingUnk2 + (float)m_fadeMs + m_timingUnk3);
+        m_startPos = start;
+        m_fadeMs = 1000;
+    }
+
+    m_stageZeroTexture = W3DAssetManager::Get_Instance()->Get_Texture(texture);
+}
+
+int WaterTracksObj::Render(DX8VertexBufferClass *vertex_buffer, int batch_start)
+{
+#ifdef BUILD_WITH_D3D8
+    Vector2 v1;
+    Vector2 v2;
+    float oo_wave_dir_len = 1.0f / m_waveDir.Length();
+    VertexFormatXYZDUV1 *verts;
+
+    if (batch_start < 4000 - m_y * m_x) {
+        if (vertex_buffer->Get_DX8_Vertex_Buffer()->Lock(vertex_buffer->FVF_Info().Get_FVF_Size() * batch_start,
+                vertex_buffer->FVF_Info().Get_FVF_Size() * m_y * m_x,
+                (BYTE **)&verts,
+                D3DLOCK_NOOVERWRITE)) {
+            return batch_start;
+        }
+    } else {
+        if (vertex_buffer->Get_DX8_Vertex_Buffer()->Lock(
+                0, vertex_buffer->FVF_Info().Get_FVF_Size() * m_y * m_x, (BYTE **)&verts, D3DLOCK_DISCARD)) {
+            return batch_start;
+        }
+
+        batch_start = 0;
+    }
+
+    float alpha;
+    float f3 = 1.0f;
+
+    if (m_type == WAVE_TYPE_UNK) {
+        v2 = m_startPos;
+        v2 -= (m_perpDir * m_length) * 0.5f;
+        v1 = v2 - ((oo_wave_dir_len * m_width) * m_waveDir);
+        alpha = 0.0f;
+
+        if (m_elapsedMs >= m_totalMs) {
+            m_elapsedMs = 0;
+        }
+
+        if (m_timingUnk1 + m_timingUnk2 - 1000.0f + (float)m_fadeMs < (float)m_elapsedMs) {
+            alpha = (float)m_elapsedMs - (m_timingUnk1 + m_timingUnk2 - 1000.0f + (float)m_fadeMs);
+            alpha = alpha / m_timingUnk3;
+            alpha = 1.0f - alpha;
+
+            if (alpha < 0.0f) {
+                alpha = 0.0f;
+            }
+        } else if (m_timingUnk1 + m_timingUnk2 - 1000.0f < (float)m_elapsedMs) {
+            alpha = (float)m_elapsedMs - (m_timingUnk1 + m_timingUnk2 - 1000.0f);
+            alpha = alpha / (float)m_fadeMs;
+
+            if (alpha > 1.0f) {
+                alpha = 1.0f;
+            }
+        }
+    } else if ((float)m_elapsedMs < m_timingUnk1) {

```

Comment:
We clearly have two different wave types here. I think structuring it as:
```
if (m_type == WAVE_TYPE_UNK){

}
else{ // Other type

}
```
would make this easier to follow.